ulid/_util.ts

```diff
@@ -0,0 +1,145 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+interface ULID {
+  (seedTime?: number): string;
+}
+
+interface PRNG {
+  (): number;
+}
+
+// These values should NEVER change. If
+// they do, we're no longer making ulids!
+export const ENCODING = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"; // Crockford's Base32
+export const ENCODING_LEN = ENCODING.length;
+export const TIME_MAX = Math.pow(2, 48) - 1;
+export const TIME_LEN = 10;
+export const RANDOM_LEN = 16;
+
+function replaceCharAt(str: string, index: number, char: string) {
+  return str.substring(0, index) + char + str.substring(index + 1);
+}
+
+export function randomChar(prng: PRNG): string {
+  let rand = Math.floor(prng() * ENCODING_LEN);
+  if (rand === ENCODING_LEN) {
+    rand = ENCODING_LEN - 1;
+  }
+  return ENCODING.charAt(rand);
+}
+
+export function encodeTime(now: number, len: number = TIME_LEN): string {
+  if (now > TIME_MAX) {
+    throw new Error("cannot encode time greater than " + TIME_MAX);
+  }
+  if (now < 0) {
+    throw new Error("time must be positive");
+  }
+  if (Number.isInteger(now) === false) {
+    throw new Error("time must be an integer");
+  }
+  let str = "";
+  for (; len > 0; len--) {
+    const mod = now % ENCODING_LEN;
+    str = ENCODING[mod] + str;
+    now = (now - mod) / ENCODING_LEN;
+  }
+  return str;
+}
+
+export function encodeRandom(len: number, prng: PRNG): string {
+  let str = "";
+  for (; len > 0; len--) {
+    str = randomChar(prng) + str;
+  }
+  return str;
+}
```

Comment:
```suggestion
export function encodeRandom(len: number): string {
  let str = "";
  const randomBytes = crypto.getRandomValues(new Uint8Array(len));
  for (let i = 0; i < len; i++) {
    str += ENCODING[randomBytes[i] % ENCODING_LEN];
  }
  return str;
}
```
This function is 10x faster, retains cryptographic randomness and removes the need for the `prng` parameter. Doing this might need tweaking elsewhere in the implementation.

Reply:
I'll run some statistical tests tomorrow, but this makes sense to me.

Reply:
This sort of breaks a lot of tests unfortunately. I'll try to pick this up on the bus tomorrow.